### PR TITLE
Filesystem Trim Support 

### PR DESCRIPTION
**Pull request by shuo-wu**
https://github.com/longhorn/longhorn/issues/836

### CODE REVIEW

**Comment by shuo-wu** [comment]
This PR is ready for review but we need to wait for the dependencies merge first.

File: controller/volume_controller.go
Comment (c3y1huang) [inline controller/volume_controller.go:1065]: Break after this?

File: types/setting.go
Comment (c3y1huang) [inline types/setting.go:872]: Why not use boolean?

File: util/util.go
Comment (shuo-wu) [reply util/util.go:912]: sorry for missing this debug log

File: controller/volume_controller.go
Comment (shuo-wu) [reply controller/volume_controller.go:1065]: We will blindly update this field for all engines and replicas.

File: controller/volume_controller.go
Comment (shuo-wu) [reply controller/volume_controller.go:1070]: All replicas should have the same value. Cannot break here.

File: types/setting.go
Comment (shuo-wu) [reply types/setting.go:866]: This determines the extra operations during the filesystem trim rather than if the trim is allowed. How about `Remove Snapshots during Filesystem Trim`?

File: types/setting.go
Comment (shuo-wu) [reply types/setting.go:872]: Nice catch. It's better to use bool here.

File: deploy/install/01-prerequisite/06-storageclass.yaml
Comment (shuo-wu) [reply deploy/install/01-prerequisite/06-storageclass.yaml:25]: No. Thank you for pointing out

File: util/util.go
Comment (derekbit) [inline util/util.go:916]: Just curious. `fstrim` is a blocking call or a non-blocking call?

File: k8s/pkg/apis/longhorn/v1beta2/volume.go
Comment (derekbit) [inline k8s/pkg/apis/longhorn/v1beta2/volume.go:74]: Should we need the empty value? How about setting the empty value to `UnmapMarkSnapChainRemovedDisabled` during upgrade?

File: engineapi/instance_manager.go
Comment (derekbit) [inline engineapi/instance_manager.go:225]: Need to check the cli version, or the volume with an old engine cannnot be started because of the `--unmap-mark-disk-chain-removed`. missing error of.

File: k8s/pkg/apis/longhorn/v1beta2/volume.go
Comment (shuo-wu) [reply k8s/pkg/apis/longhorn/v1beta2/volume.go:74]: This will be updated `UnmapMarkSnapChainRemoved("ignored")`.

File: util/util.go
Comment (shuo-wu) [reply util/util.go:916]: I think it's a blocking call. I need to have a test (by deleting tons of small files) for this part.

Comment (shuo-wu) [reply util/util.go:916]: I tried to create 50000 4MB files (200G in total) in a 500G volume then delete and trim them, `fstrim` gets stuck for around 2 seconds. I don't think there is a timeout risk.

**Comment by innobead** [comment]
@shuo-wu Is this ready for review?

**Comment by shuo-wu** [comment]
Yes. We need to wait for the engine im PRs before merging this.

File: datastore/longhorn.go
Comment (innobead) [inline datastore/longhorn.go:816]: Why filtering DR volumes only?

File: datastore/longhorn.go
Comment (innobead) [inline datastore/longhorn.go:802]: Where/when has `LonghornLabelFollowGlobalSetting`  been updated to volumes?

**Review by innobead — COMMENTED** [review]
In general, LGTM, but have question about `ListVolumesFollowsGlobalSettingsRO`.

File: datastore/longhorn.go
Comment (shuo-wu) [reply datastore/longhorn.go:816]: This is typo. Removed the check.

File: datastore/longhorn.go
Comment (shuo-wu) [reply datastore/longhorn.go:802]: Missed the changes in the mutation webhook part.

**Comment by shuo-wu** [comment]
> In general, LGTM, but have question about `ListVolumesFollowsGlobalSettingsRO`.

I just update the commit `VolumeController enqueue Setting RemoveSnapshotsDuringFilesystemTrim` for this issue. PTAL.

File: util/util.go
Comment (innobead) [inline util/util.go:910]: nit: Suggest having a meaning parameter name like `volumeName`

File: go.mod
Comment (innobead) [inline go.mod:7]: Wrong dependency

File: k8s/crds.yaml
Comment (innobead) [inline k8s/crds.yaml:1364]: Why there is no `UnmapMarkSnapChainRemovedEnabled` in the engine status?

File: webhook/resources/volume/mutator.go
Comment (innobead) [inline webhook/resources/volume/mutator.go:110]: Can use `UnmapMarkSnapChainRemovedIgnored` from v1beta2 volume instead of text `ignored` here

File: webhook/resources/volume/mutator.go
Comment (innobead) [inline webhook/resources/volume/mutator.go:247]: Use `UnmapMarkSnapChainRemovedIgnored` instead of `ignored`

File: engineapi/instance_manager.go
Comment (innobead) [inline engineapi/instance_manager.go:218]: This part needs to rebase to do backward compatibility support

**Review by innobead — CHANGES_REQUESTED** [review]

File: k8s/crds.yaml
Comment (shuo-wu) [reply k8s/crds.yaml:1364]: Sorry. I just manually updated the CRD in repo `longhorn` and forgot to update this.

File: engineapi/instance_manager.go
Comment (shuo-wu) [reply engineapi/instance_manager.go:218]: It seems that we need to check the backward compatibility for the below option `dataLocality == longhorn.DataLocalityStrictLocal` as well. I will directly include it in this commit.

File: k8s/crds.yaml
Comment (innobead) [reply k8s/crds.yaml:1364]: I see, but next time you can use https://github.com/longhorn/longhorn-manager/blob/master/k8s/generate_code.sh#L70 to update the crds.yaml directly.

File: controller/volume_controller.go
Comment (innobead) [inline controller/volume_controller.go:1053]: Suggest renaming to `isUnmapMarkSnapChainRemovedEnabled`

File: webhook/resources/volume/mutator.go
Comment (innobead) [inline webhook/resources/volume/mutator.go:187]: `SettingNameReplicaAutoBalance` is before 1.4, so do we need to do the upgrade migration?

**Review by innobead — CHANGES_REQUESTED** [review]
LGTM, except a few comments needed to resolve.

File: types/types.go
Comment (innobead) [inline types/types.go:207]: This is good.

File: webhook/resources/volume/mutator.go
Comment (shuo-wu) [reply webhook/resources/volume/mutator.go:187]: I didn't add an upgrade path for this. This mutator blindly checks the labels during update/creation.
And this labeling makes Longhorn update the volume/engine fields in time when the related settings are modified.

Comment (innobead) [reply webhook/resources/volume/mutator.go:187]: I see, it's OK.

**Review by innobead — APPROVED** [review]

**Pull request merged**
